pkg/server/apiextensions.go

```diff
@@ -245,117 +259,226 @@ func (c *apiBindingAwareCRDLister) ListWithContext(ctx context.Context, selector
-				return nil, err
-			}
-		}
+	if strings.Contains(name, "kcp.dev") {
+		// Priority 1: system CRD
+		crd, err = c.getSystemCRD(clusterName, name)
+	} else if identity := IdentityFromContext(ctx); identity != "" {
+		// Priority 2: identity request
+		crd, err = c.getForIdentity(name, identity)
+	} else if isPartialMetadataRequest(ctx) {
+		// Priority 3: partial metadata
+		crd, err = c.getForPartialMetadata(name)
+	} else {
+		// Priority 4: full data
+		crd, err = c.getForFullData(clusterName, name)
+	}
 
-		if crd == nil {
-			return nil, apierrors.NewNotFound(schema.GroupResource{Group: apiextensionsv1.SchemeGroupVersion.Group, Resource: "customresourcedefinitions"}, name)
-		}
+	if err != nil {
+		return nil, err
+	}
 
-		return crd, nil
+	if isPartialMetadataRequest(ctx) {
+		crd = partialMetadataCRD(crd)
 	}
 
-	// Priority 1: see in system CRDs
-	systemCRDKeys := c.systemCRDProvider.Keys(ctx, cluster.Name)
-	systemCRDKeyName := clusters.ToClusterAwareKey(SystemCRDLogicalCluster, name)
-	if systemCRDKeys.Has(systemCRDKeyName) {
-		crd, err = c.crdLister.Get(systemCRDKeyName)
-		if err != nil && !apierrors.IsNotFound(err) {
-			// something went wrong w/the lister - could only happen if meta.Accessor() fails on an item in the store.
-			return nil, err
+	return crd, nil
+}
+
+func partialMetadataCRD(crd *apiextensionsv1.CustomResourceDefinition) *apiextensionsv1.CustomResourceDefinition {
+	crd = crd.DeepCopy()
+
+	if crd.Annotations == nil {
+		crd.Annotations = make(map[string]string)
+	}
+
+	crd.Annotations[annotationKeyPartialMetadata] = ""
+
+	// set minimal schema that prunes everything but ObjectMeta
+	for _, v := range crd.Spec.Versions {
+		v.Schema = &apiextensionsv1.CustomResourceValidation{
+			OpenAPIV3Schema: &apiextensionsv1.JSONSchemaProps{
+				Type: "object",
+			},
 		}
+	}
+
+	return crd
+}
+
+func (c *APIBindingAwareCRDLister) getForFullData(clusterName logicalcluster.LogicalCluster, name string) (*apiextensionsv1.CustomResourceDefinition, error) {
+	if clusterName == logicalcluster.Wildcard {
+		return c.getWildcard(name)
+	}
+
+	return c.get(clusterName, name)
+}
+
+func (c *APIBindingAwareCRDLister) getWildcard(name string) (*apiextensionsv1.CustomResourceDefinition, error) {
+	// TODO use an index by group+resource
+	crds, err := c.crdLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	crd, equivalentSchemas := findCRD(name, crds)
+	if !equivalentSchemas {
+		err = apierrors.NewInternalError(fmt.Errorf("error resolving resource: cannot watch across logical clusters for a resource type with several distinct schemas"))
+		return nil, err
+	}
+
+	if crd == nil {
+		return nil, errors.NewNotFound(apiextensionsv1.Resource("customresourcedefinitions"), name)
+	}
+
+	return crd, nil
+}
+
+func (c *APIBindingAwareCRDLister) getForIdentity(name, identity string) (*apiextensionsv1.CustomResourceDefinition, error) {
+	apiExports, err := c.apiExportIndexer.ByIndex("byIdentity", identity)
```

Comment:
this confuses me. Why do you need the export in the first place? What is missing in the binding to be sufficient here?

Reply:
This is only for `/clusters/*/apis/$group/$version/$resource:identity`. We don't know the binding.